content/en/blog/_posts/2021-08-11-memory-qos-cgroups-v2.md

```diff
@@ -0,0 +1,73 @@
+---
+layout: blog
+title: 'Support memory qos with cgroups v2'
+date: 2021-08-11
+slug: memory-qos-cgroups-v2
+---
+
+**Authors:** Tim Xu (Tencent Cloud)
+
+Memory QoS is used to memory protection and usage throttle on pod / container to increase memory isolation and service quality.
+
+In traditional cgroups v1 in Kubernetes, cpu resources can be very well protected. Kubernetes uses `cpu_shares / cpu_set / cpu_quota / cpu_period` to isolate cpu. But the memory is not well guaranteed, especially at cgroup level. Kubernetes only uses `memory.limit_in_bytes=sum(pod.spec.containers.resources.limits[memory])` with cgroups v1 and `memory.max=sum(pod.spec.containers.resources.limits[memory])` with cgroups v2 to limit memory usage. `resources.requests[memory]` is not yet used neither by cgroups v1 nor cgroups v2 to protect memory requests. About memory protection, we use `oom_scores` to determine order of killing container process when OOM occurs.
+
+It may cause:
+- Pod / Container memory requests can't be fully reserved, page cache is at risk of being recycled
+- Pod / Container memory allocation is not well protected, there may occur allocation latency frequently when node memory nearly runs out
```

Comment:
```suggestion
- Pod or Container memory allocation to not be well protected and allocation latency frequently may occur when node memory nearly runs out
```

Reply:
I'm not sure I understand what this means so I can't work out a good way to paraphrase it. What form of allocation is being discussed here? `sbrk()` and friends?